s3stream/src/main/java/com/automq/stream/s3/wal/impl/s3/ObjectStorageWALService.java

```diff
@@ -0,0 +1,148 @@
+import com.automq.stream.s3.wal.util.WALUtil;
+import com.automq.stream.utils.Time;
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.CompositeByteBuf;
+import io.netty.buffer.PooledByteBufAllocator;
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.automq.stream.s3.wal.common.RecordHeader.RECORD_HEADER_MAGIC_CODE;
+import static com.automq.stream.s3.wal.common.RecordHeader.RECORD_HEADER_SIZE;
+
+public class ObjectStorageWALService implements WriteAheadLog {
+    private static final Logger log = LoggerFactory.getLogger(ObjectStorageWALService.class);
+
+    protected ObjectStorage objectStorage;
+    protected RecordAccumulator accumulator;
+
+    public ObjectStorageWALService(Time time, ObjectStorage objectStorage, ObjectStorageWALConfig config) {
+        this.objectStorage = objectStorage;
+        this.accumulator = new RecordAccumulator(time, objectStorage, config);
+    }
+
+    @Override
+    public WriteAheadLog start() throws IOException {
+        log.info("Start S3 WAL.");
+        accumulator.start();
+        return this;
+    }
+
+    @Override
+    public void shutdownGracefully() {
+        log.info("Shutdown S3 WAL.");
+        accumulator.close();
+    }
+
+    @Override
+    public WALMetadata metadata() {
+        return null;
+    }
+
+    private ByteBuf recordHeader(ByteBuf body, int crc, long start) {
+        return new RecordHeader()
+            .setMagicCode(RECORD_HEADER_MAGIC_CODE)
+            .setRecordBodyLength(body.readableBytes())
+            .setRecordBodyOffset(start + RECORD_HEADER_SIZE)
+            .setRecordBodyCRC(crc)
+            .marshal();
+    }
+
+    private ByteBuf record(ByteBuf body, int crc, long start) {
+        CompositeByteBuf record = ByteBufAlloc.compositeByteBuffer();
+        crc = 0 == crc ? WALUtil.crc32(body) : crc;
+        record.addComponents(true, recordHeader(body, crc, start), body);
+        return record;
+    }
+
+    @Override
+    public AppendResult append(TraceContext context, ByteBuf data, int crc) throws OverCapacityException {
+        final long recordSize = RECORD_HEADER_SIZE + data.readableBytes();
+        final CompletableFuture<AppendResult.CallbackResult> appendResultFuture = new CompletableFuture<>();
+        long expectedWriteOffset = accumulator.append(recordSize, start -> record(data, crc, start), appendResultFuture);
+
+        return new AppendResultImpl(expectedWriteOffset, appendResultFuture);
+    }
+
+    @Override
+    public Iterator<RecoverResult> recover() {
+        return new Iterator<>() {
+            // TODO: Recover only the objects with the latest epoch.
+            private List<RecordAccumulator.WALObject> objectList = accumulator.objectList();
+            private int nextIndex = 0;
+            private ByteBuf record = PooledByteBufAllocator.DEFAULT.buffer();
```

Comment:
We should use `ByteBufAlloc` to allocate buffer, in the package `com.automq.stream.s3`